extensions/websocket.md

```diff
@@ -0,0 +1,38 @@
+---
+title: IRCv3 websocket support
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: Darren Whitlen
+    period: 2018
+    email: darren@kiwiirc.com
+---
+
+## Introduction
+
+Web based IRC clients are becoming more common and they all increase the demand for websocket connectivity to IRC networks.
+
+This specification provides a common implementation for web and other websocket clients to connect to IRC networks using common and standard development tools used on web pages, with a focus on a simple client facing interface.
+
+## Websocket features and encoding
+
+Websockets offer different transports of communication, binary and strings. With the consideration that the IRC protocol is entirely lines of text, we make use of the message based text implementation that websockets natively provide. These are UTf8 encoded text messages where one websocket message consists of a single complete IRC message line.
+
+The internal websocket protocol handles message fragmentation and termination via frames, which is out of scope for this document.
+
+Because web browsers are pushing more towards a UTF8 default encoding for HTML5 documents and scriptings languages, and these web browsers are the primary use case for websockets, IRC servers MUST encode messages as UTF8 before sending them to the client. Ignoring this may result in the browser closing the websocket connection on invalid UTF8 sequences.
```

Comment:
> IRC servers MUST encode messages as UTF8 before sending them to the client.

An IRC server can determine that an incoming message isn't valid UTF-8, but how is it supposed to know the actual original encoding to translate from?

IIRC this was the primary argument for using binary mode with IRC, rather than text mode. (It seems that's what [InspIRCd already does][1].)

[1]: https://github.com/inspircd/inspircd/blob/master/src/modules/m_websocket.cpp#L356

Reply:
I would expect a configuration option for fallback encoding, because most encodings are impossible to detect.

Or possibly the option to not detect the encoding and just filtering/replacing the illegal characters

Reply:
Can't we just ban non-UTF-8 from the IRC protocol once and for all?

Reply:
I have a new IRC friend who wants to connect to modern IRC using old computer hardware, including the Apple II, DOS-era machines, etc.

Detection / negotiation are a different issue, but IMO non-UTF-8 support shouldn't be deprecated.

(I know granted they aren't likely to choose a WebSocket transport for their Apple II client, but still, on the larger topic...)